net/tsdial/dnsmap.go

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) 2021 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package tsdial
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strconv"
+	"strings"
+
+	"inet.af/netaddr"
+	"tailscale.com/types/netmap"
+	"tailscale.com/util/dnsname"
+)
+
+// DNSMap maps MagicDNS names (both base + FQDN) to their first IP.
+// It should not be mutated once created.
+type DNSMap map[string]netaddr.IP
```

Comment:
use dnsname.FQDN as the key? Forces running dnsname.ToFQDN to populate, which eliminates issues around trailing dot vs. no trailing dot.

Reply:
well, per its docs, it holds bare names too, without FQDNs, so you can `curl http://foo` over SOCKS/HTTP proxy.

So dsnname.FQDN would be confusing in a different way. I can list some example keys formats, though.

Reply:
(also, this is just code movement so I'm reluctant to do too much rewriting during a git mv... but we can later)